[core] out of band serialization exception (#47544)
ray-project/ray: doc/source/ray-core/patterns/out-of-band-object-ref-serialization.rst

.. _ray-out-of-band-object-ref-serialization:

> Anti-pattern: Serialize ray.ObjectRef out of band

Reviewer: cross-reference it in https://docs.ray.io/en/latest/ray-core/objects/serialization.html#serializing-objectrefs

=================================================

> **TLDR:** Avoid serialize ``ray.ObjectRef`` because Ray cannot know when the GC the underlying object.

Reviewer 2: ```suggestion
**TLDR:** Avoid serializing ``ray.ObjectRef`` because Ray can't know when to garbage collect the underlying object.
```

> Ray's ``ray.ObjectRef`` is distributed reference counted. Ray pins the underlying object until the reference is not used by the system anymore.

Reviewer 2: ```suggestion
Ray's ``ray.ObjectRef`` is distributed reference counted. Ray pins the underlying object until the reference isn't used by the system anymore.
```

> When all references are gone, the pinned object is garbage collected and cleaned up from the system.

Reviewer 2: ```suggestion
When all references are gone, Ray garbage collects the pinned object and cleans it up from the system.
```

> However, if user code serializes ``ray.objectRef``, Ray cannot keep track of the reference.

Reviewer 2: ```suggestion
However, if user code serializes ``ray.objectRef``, Ray can't keep track of the reference.
```

> To avoid incorrect behavior, if ``ray.ObjectRef`` is serialized by ``ray.cloudpickle``, Ray pins the object for the lifetime of a worker. "pin" means that object cannot be evicted from the object store

Reviewer 2: ```suggestion
To avoid incorrect behavior, if ``ray.cloudpickle`` serializes``ray.ObjectRef``, Ray pins the object for the lifetime of a worker. "Pin" means that object can't be evicted from the object store
```

> until the corresponding worker dies. It is prone to Ray object leaks, which can lead disk spilling.

Reviewer: ```suggestion
until the corresponding owner worker dies. It is prone to Ray object leaks, which can lead disk spilling.
```

To detect if this pattern exists in your code, you can set an environment variable ``RAY_allow_out_of_band_object_ref_serialization=0``. If Ray detects

> ``ray.ObjectRef`` is serialized by ``ray.cloudpickle``, it raises an exception with helpful messages.

Reviewer 2: ```suggestion
that ``ray.cloudpickle`` serialized``ray.ObjectRef``, it raises an exception with helpful messages.
```
